### ISSUE TITLE
AE-PSNR and LR-PSNR scores

### ISSUE THREAD

**Issue by tomcan2015** [2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution#9]
In your Supplementary Material of the paper, you mentioned AE-PSNR and LR-PSNR as the fidelity measurement. My understanding is the AE-PSNR compares the AE output with HR and LR-PSNR compares the AE output with LR. Am I right? I am training a 2x SR network, should I retrain your AE_RRDBdecoder_100K.pth since I am using different scaling ratio?

**Comment by 2minkyulee** [comment]
1. AE-PSNR measures PSNR btw AE(HR) and AE(SR).
2. LR-PSNR measures PSNR btw downsample(HR) and downsample(SR)
3. When measuring AE-PSNR for x2 SR task, I recommend retraining a new AE, with a bottleneck that has x0.5 resolution and 3 channel.

**Comment by tomcan2015** [comment]
Thanks for your quick response. I am training a SR 2x network to restore HR from LR. It is not a simple bicubic down scaled LR, the downscaled LR may be compressed or not. So it is similar to the “real” case but I don’t consider blur, noise, only compression artifacts (if any) need be removed from LR.
As you suggested, I am starting to train AE first for 2x. I have two options here:
1. Use bicubic down scaled clean LR without compression to train AE, both encoder and decoder. This is similar to the setting in [train_Synthetic_AE_RRDB_LRrecon1](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Synthetic_AE_RRDB_LRrecon1.yml)
2. Train the decoder first with compressed LR (actually mixed compressed and uncompressed), Then freeze the decoder and train encoder with clean LR without compression. This is similar to the setting in [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml)
In option2, I don’t have the “RealESRNet_x4plus.pth  # official RealESRNet Weight”, so I have to train the decoder first.
Which option should I use?
There is another option 3: train the encoder and decoder together with compressed LR (actually mixed compressed with uncompressed). This might the best way to let AE learn the optimal fidelity point based on the input data.
Your guidance is appreciated.

**Comment by 2minkyulee** [comment]
I suggest below, since the **AutoEncoding should approximate your SR task (fidelity version)**.

1. pretrain real-world (or your SR setting) x2 **fidelity** SR.
2. Train a AE with 1) the pretrained SR network as a decoder, and 2) a regularization loss that makes the bottleneck approximate your LR.
3. Train your **perceptual** SR, with your customized AESOP loss (here, freeze AE).

**Comment by tomcan2015** [comment]
Thanks for your suggestion. In step 2 of your suggestion above, I need to use real image (mixed compressed and uncompressed) to minimize both l_pix and l_pix_at_LR?
How about my option3 to train both encoder and decoder with real image the same time?

**Comment by 2minkyulee** [comment]
Regarding my suggestion, l_pix takes HR images (so they should be clean images). 
l_pix_at_LR should take mixed compressed LR image.

Regarding your option 3, I believe that it will work too.
My suggestion has a benefit that your overall training will be faster (since we reuse the fidelity SR net for both 1) perceptual SR training and 2) AE training).

**Comment by tomcan2015** [comment]
Thanks for your comment. I don’t quite understand the last sentence about reuse the fidelity SR net. Here I am using the RRDB net trained with fidelity (only L1 loss).
For perceptual SR training (my SR net, let me call it CNN net), I will freeze the AE and use AESOP loss from AE to train CNN. This step is the same for option 2 and 3.
The only difference between option2 and 3 is: option2 train decoder (RRDB) and encoder separately. Option3 train decoder and encoder together. It seems option3 will be faster?
Your AE training for real image is option2, but your used the pretrained “RealESRNet_x4plus.pth # official RealESRNet Weight” in [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml)
For me, I don’t have such x2 checkpoint for RRDB so I have to train RRDB with mixed compressed image to get such a decoder.
After I get the x2 decoder, I will freeze it and train encoder with mixed image. There are two loss functions need be minimized :l_pix takes HR images (clean images),l_pix_at_LR takes mixed compressed LR image.
In your [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml) setting, it seems you used clean LR instead of real LR with degradations to calculate l_pix_at_LR.

**Comment by tomcan2015** [comment]
Let me try to explain what I am confused:
In your example: [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml)
You are using the RealESRNet_x4plus.pth, this is a perceptual SR!
Then you freeze the decoder and train encoder with clean LR to get AE.
But what you suggested me is : train RRDB with fidelity as decoder.
Then freeze decoder and train AE with mixed compressed LR.
This method is different with the example: [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml)

**Comment by 2minkyulee** [comment]
RealESRNet is a fidelity SR network; RealESR**GAN** is the perceptual SR network.
The AutoEncoder should be trained purely on regression (e.g., only L1 or L2 loss).

In more details, RealESRNet is finetuned from ESR**GAN**, so it may seem as a perceptual SR network. However, my view is that since RealESRNet training only uses L1 loss, it is fundamentally a fidelity SR network.

I'm not sure why the authors have finetuned from ESRGAN instead of RRDB, but I believe that this is simply an empirical choice that worked better.

**Comment by 2minkyulee** [comment]
> My suggestion has a benefit that your overall training will be faster (since we reuse the fidelity SR net for both 1) perceptual SR training and 2) AE training).

Regarding above, I'm considering the SRGAN-like training framework where we train perceptual SR networks given a pretrained fidelity SR network. 

In this situation, both the decoder (of AE) and the perceptual SR net can use a single pretrained fidelity SR network weight (i.e., dont need to train AE from scratch).

**Comment by 2minkyulee** [comment]
> In your [train_Realworld_AE_RRDB_DecoderFreeze.yml](https://github.com/2minkyulee/AESOP-Auto-Encoded-Supervision-for-Perceptual-Image-Super-Resolution/blob/main/AESOP/options/train/AutoEncoder/train_Realworld_AE_RRDB_DecoderFreeze.yml) setting, it seems you used clean LR instead of real LR with degradations to calculate l_pix_at_LR.

Regarding above, your concerns are valid. The current version (train_Realworld_AE_RRDB_DecoderFreeze.yml; which aligns with your option 2) was a choice for implementation simplicity; and may not be that accurate. (But I believe that this is a sufficiently fair approximation, and wont have significant issues.)

The suggestions I have gave you in this issue thread (using compressed LR for bottleneck loss) should be the more "accurate" way.

**Comment by tomcan2015** [comment]
Thanks for your detailed answers! Now I am clean about what I should do. I will follow your suggestion.
The key point is to let AE output only the fidelity part of the SR output based on the input data and don’t conflict with the perceptual part of loss functions. This is a genius idea! I am wondering why so many researchers didn’t notice that? Your paper is a milestone!

**Comment by 2minkyulee** [comment]
Thank you for you acknowledgement :)
